parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,413 @@
+    final double MAX = ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES << 3;
+    int numBits = (int)M;
+
+    // Handle overflow.
+    if (M > MAX || M < 0) {
+      numBits = (int)MAX;
+    }
+
+    // Get next power of 2 if bits is not power of 2.
+    if ((numBits & (numBits - 1)) != 0) {
+      numBits = Integer.highestOneBit(numBits) << 1;
+    }
+
+    // Minimum
+    if (numBits < (BYTES_PER_BUCKET << 3)) {
+      numBits = BYTES_PER_BUCKET << 3;
+    }
+
+    return numBits;
+  }
+
+  /**
+   * used to decide if we want to work to the next page
+   * @return Bytes buffered of bloom filter.
+   */
+  public long getBufferedSize() {
+    return bitset.length;
+  }
+
+  /**
+   * Compute hash for int value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(int value) {
+    ByteBuffer plain = ByteBuffer.allocate(Integer.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putInt(value);
+    switch (hash) {
+      case MURMUR3_X64_128: return Murmur3.hash64(plain.array());
+      default:
+        throw new RuntimeException("Not support hash strategy");
+    }
+  }
+
+  /**
+   * Compute hash for long value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(long value) {
+    ByteBuffer plain = ByteBuffer.allocate(Long.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putLong(value);
+    switch (hash) {
+      case MURMUR3_X64_128: return Murmur3.hash64(plain.array());
+      default:
+        throw new RuntimeException("Not support hash strategy");
+    }
+  }
+
+  /**
+   * Compute hash for double value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(double value) {
+    ByteBuffer plain = ByteBuffer.allocate(Double.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putDouble(value);
+    switch (hash) {
+      case MURMUR3_X64_128: return Murmur3.hash64(plain.array());
+      default:
+        throw new RuntimeException("Not support hash strategy");
+    }
+  }
+
+  /**
+   * Compute hash for float value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(float value) {
+    ByteBuffer plain = ByteBuffer.allocate(Float.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putFloat(value);
+    switch (hash) {
+      case MURMUR3_X64_128: return Murmur3.hash64(plain.array());
+      default:
+        throw new RuntimeException("Not support hash strategy");
+    }
+  }
+
+  /**
+   * Compute hash for Binary value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(Binary value) {
+    try {
+      ByteBuffer plain = ByteBuffer.allocate(Integer.SIZE/Byte.SIZE);
+      plain.order(ByteOrder.LITTLE_ENDIAN).putInt(value.length());
```

Comment:
Why does the hash include the length of the string? That seems unnecessary.

Reply:
Some hash functions (I don't know if Murmur3 is one of them) have trivial collisions when length is not included in the hashing

Reply:
You mean something like "abc" and "abc\x00\x00" hash to the same value?

I haven't seen any warnings about this behavior in murmur3 implementations and a quick google search doesn't turn up much. I'd prefer not to include the length because it makes the hash harder to compute, but we should either make sure we're okay with the trivial collisions or find out whether this is true.

I'm leaning toward just accepting possible collisions, depending on the definition of trivial. We expect low cardinality columns to be dictionary-encoded. High cardinality columns would be more like UUIDs or other sufficiently random values that wouldn't hit this case very much.

Reply:
This doesn't affect murmur3. The length of the buffer is XORed into the result: https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp#L141

Reply:
We defined to compute plain encoding of value. So it depends on how plain encoding of binary value it is. From the PlainValueWriter.java, it store the length of Binary.